internal/security/secretstore/tokenfilewriter/tokenfilewriter.go

```diff
@@ -0,0 +1,172 @@
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/secretstore/secretsengine"
+	"github.com/edgexfoundry/edgex-go/internal/security/secretstore/tokencreatable"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
+	"github.com/edgexfoundry/go-mod-secrets/v2/pkg/token/fileioperformer"
+	"github.com/edgexfoundry/go-mod-secrets/v2/secrets"
+)
+
+const (
+	consulSecretsEngineOpsPolicyName = "consul_secrets_engine_management_policy"
+)
+
+// TokenFileWriter is a mechanism to generates a token and writes it into a file specified by configuration
+type TokenFileWriter struct {
+	logClient    logger.LoggingClient
+	secretClient secrets.SecretStoreClient
+	fileOpener   fileioperformer.FileIoPerformer
+}
+
+// NewWriter instantiates a TokenFileWriter instance
+func NewWriter(lc logger.LoggingClient,
+	sc secrets.SecretStoreClient,
+	fileOpener fileioperformer.FileIoPerformer) TokenFileWriter {
+	return TokenFileWriter{
+		logClient:    lc,
+		secretClient: sc,
+		fileOpener:   fileOpener,
+	}
+}
+
+// CreateAndWrite generates a new token and writes it to the file specified by tokenFilePath
+// the generation of the token requires root token privilege
+// it overwrites the file if already exists
+// returns error if anything fails during the whole process
+func (w TokenFileWriter) CreateAndWrite(rootToken string, tokenFilePath string,
+	createTokenFunc tokencreatable.CreateTokenFunc) (tokencreatable.RevokeFunc, error) {
+	if len(rootToken) == 0 {
+		return nil, fmt.Errorf("rootToken is required")
+	}
+
+	// creates the token
+	createTokenFuncName := getFunctionName(createTokenFunc)
+	tokenCreated, revokeTokenFunc, createErr := createTokenFunc(rootToken)
+	if createErr != nil {
+		return nil, fmt.Errorf("failed to create token with %s: %s", createTokenFuncName, createErr.Error())
+	}
+
+	w.logClient.Infof("created token with %s", createTokenFuncName)
+
+	var fileErr error
+	defer func() {
+		// call revokeTokenFunc if there is some fileErr and revokeTokenFunc itself is not nil
+		if fileErr != nil && revokeTokenFunc != nil {
+			revokeTokenFunc()
+		}
+	}()
+
+	// Write the token issuing token to disk to pass it to the token provider
+	tokenFileAbsPath, fileErr := filepath.Abs(tokenFilePath)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to convert tokenFile to absolute path %s: %s", tokenFilePath, fileErr.Error())
+	}
+
+	dirOfCreatedToken := filepath.Dir(tokenFileAbsPath)
+	fileErr = w.fileOpener.MkdirAll(dirOfCreatedToken, 0700)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create tokenpath base dir: %s", fileErr.Error())
+	}
+
+	fileWriter, fileErr := w.fileOpener.OpenFileWriter(tokenFileAbsPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create token file %s: %s", tokenFileAbsPath, fileErr.Error())
+	}
+
+	if fileErr = json.NewEncoder(fileWriter).Encode(tokenCreated); fileErr != nil {
```

Comment:
This line can fail if NewEncoder() returns nil.

Reply:
could not think of any case, could you give me a concrete example that json.NewEncoder will fail?

Reply:
@bnevis-i :   It is impossible for `json.NewEncoder` to return `nil`, please check out the golang source code at https://golang.org/src/encoding/json/stream.go  line 191-194, in which it is always return an instance of `&Encoder{w: w, escapeHTML: true}`.  Thus `json.NewEncoder()` will never return nil.

Reply:
Good to know.